File: backend/utils/otp.py
import random
import string
from datetime import datetime, timedelta
from flask_mail import Message
from twilio.rest import Client
from config import Config
from app import mail, db
from models import OTP  # We'll need to add this model

class OTPManager:
    def __init__(self):
        self.mail = mail
        self.twilio_client = Client(Config.TWILIO_ACCOUNT_SID, Config.TWILIO_AUTH_TOKEN)
        self.otp_expiry_minutes = 5  # OTP validity period

    def generate_otp(self, length=6):
        """Generate a random numeric OTP"""
        return ''.join(random.choice(string.digits) for _ in range(length))

    def send_email_otp(self, email):
        """Send OTP via email"""
        otp_code = self.generate_otp()
        expiration = datetime.utcnow() + timedelta(minutes=self.otp_expiry_minutes)

        # Store in database
        otp_record = OTP(
            recipient=email,
            otp_code=otp_code,
            otp_type='email',
            expires_at=expiration
        )
        db.session.add(otp_record)
        db.session.commit()

        # Send email
        msg = Message(
            subject="Your Verification Code",
            sender=Config.MAIL_USERNAME,
            recipients=[email]
        )
        msg.body = f"Your verification code is: {otp_code}\nThis code expires in {self.otp_expiry_minutes} minutes."
        
        try:
            self.mail.send(msg)
            return True
        except Exception as e:
            db.session.delete(otp_record)
            db.session.commit()
            return False

    def send_sms_otp(self, phone_number):
        """Send OTP via SMS using Twilio"""
        if not Config.TWILIO_ACCOUNT_SID:
            return False

        otp_code = self.generate_otp()
        expiration = datetime.utcnow() + timedelta(minutes=self.otp_expiry_minutes)

        # Store in database
        otp_record = OTP(
            recipient=phone_number,
            otp_code=otp_code,
            otp_type='sms',
            expires_at=expiration
        )
        db.session.add(otp_record)
        db.session.commit()

        # Send SMS
        try:
            message = self.twilio_client.messages.create(
                body=f"Your verification code is: {otp_code}",
                from_=Config.TWILIO_PHONE_NUMBER,
                to=phone_number
            )
            return True
        except Exception as e:
            db.session.delete(otp_record)
            db.session.commit()
            return False

    def verify_otp(self, recipient, otp_code):
        """Verify OTP from database"""
        otp_record = OTP.query.filter_by(
            recipient=recipient,
            otp_code=otp_code
        ).first()

        if not otp_record:
            return False

        if otp_record.expires_at < datetime.utcnow():
            db.session.delete(otp_record)
            db.session.commit()
            return False

        # Clean up used OTP
        db.session.delete(otp_record)
        db.session.commit()
        return True

    def cleanup_expired_otps(self):
        """Clean up expired OTPs from database"""
        expired = OTP.query.filter(OTP.expires_at < datetime.utcnow()).all()
        for otp in expired:
            db.session.delete(otp)
        db.session.commit()
        return len(expired)

File: backend/utils/otp.js
const nodemailer = require('nodemailer');
const twilio = require('twilio');
const config = require('../config');
const { OTP } = require('../models');

class OTPManager {
  constructor() {
    this.otpExpiryMinutes = config.OTP_EXPIRY_MINUTES;
    
    // Initialize email transporter
    if (config.MAIL_USERNAME && config.MAIL_PASSWORD) {
      this.emailTransporter = nodemailer.createTransporter({
        host: config.MAIL_HOST,
        port: config.MAIL_PORT,
        secure: false,
        auth: {
          user: config.MAIL_USERNAME,
          pass: config.MAIL_PASSWORD
        }
      });
    }
    
    // Initialize Twilio client
    if (config.TWILIO_ACCOUNT_SID && config.TWILIO_AUTH_TOKEN) {
      this.twilioClient = twilio(config.TWILIO_ACCOUNT_SID, config.TWILIO_AUTH_TOKEN);
    }
  }

  generateOTP(length = 6) {
    return Math.random().toString().substr(2, length);
  }

  async sendEmailOTP(email) {
    if (!this.emailTransporter) {
      throw new Error('Email configuration not set up');
    }

    const otpCode = this.generateOTP();
    const expiration = new Date(Date.now() + this.otpExpiryMinutes * 60 * 1000);

    try {
      // Store in database
      await OTP.create({
        recipient: email,
        otp_code: otpCode,
        otp_type: 'email',
        expires_at: expiration
      });

      // Send email
      const mailOptions = {
        from: config.MAIL_USERNAME,
        to: email,
        subject: 'Your Verification Code',
        text: `Your verification code is: ${otpCode}\nThis code expires in ${this.otpExpiryMinutes} minutes.`
      };

      await this.emailTransporter.sendMail(mailOptions);
      return true;
    } catch (error) {
      console.error('Email OTP error:', error);
      // Clean up failed OTP record
      await OTP.destroy({ where: { recipient: email, otp_code: otpCode } });
      return false;
    }
  }

  async sendSMSOTP(phoneNumber) {
    if (!this.twilioClient) {
      throw new Error('Twilio configuration not set up');
    }

    const otpCode = this.generateOTP();
    const expiration = new Date(Date.now() + this.otpExpiryMinutes * 60 * 1000);

    try {
      // Store in database
      await OTP.create({
        recipient: phoneNumber,
        otp_code: otpCode,
        otp_type: 'sms',
        expires_at: expiration
      });

      // Send SMS
      await this.twilioClient.messages.create({
        body: `Your verification code is: ${otpCode}`,
        from: config.TWILIO_PHONE_NUMBER,
        to: phoneNumber
      });

      return true;
    } catch (error) {
      console.error('SMS OTP error:', error);
      // Clean up failed OTP record
      await OTP.destroy({ where: { recipient: phoneNumber, otp_code: otpCode } });
      return false;
    }
  }

  async verifyOTP(recipient, otpCode) {
    try {
      const otpRecord = await OTP.findOne({
        where: {
          recipient: recipient,
          otp_code: otpCode
        }
      });

      if (!otpRecord) {
        return false;
      }

      if (otpRecord.expires_at < new Date()) {
        await OTP.destroy({ where: { id: otpRecord.id } });
        return false;
      }

      // Clean up used OTP
      await OTP.destroy({ where: { id: otpRecord.id } });
      return true;
    } catch (error) {
      console.error('OTP verification error:', error);
      return false;
    }
  }

  async cleanupExpiredOTPs() {
    try {
      const expiredCount = await OTP.destroy({
        where: {
          expires_at: {
            [require('sequelize').Op.lt]: new Date()
          }
        }
      });
      return expiredCount;
    } catch (error) {
      console.error('Cleanup expired OTPs error:', error);
      return 0;
    }
  }
}

module.exports = new OTPManager(); 